docs/design/auditing-containment.md

```diff
@@ -0,0 +1,168 @@
+```
+However, we're departing from the whitepaper because offline nodes will not be moved to containment mode.
+They will just be marked as offline.
+It's only when nodes initially respond to the audit service's dial but then refuse to send the requested erasure share that they are moved to containment mode.
+
+The node will be given a `contained` flag, and if the audit service attempts to audit the node again, it will see this flag and request the same share it had originally requested.
+If the storage node fails to send this share again, it will have its `reverifyCount` incremented.
+
+If the `failedReverifyCount` exceeds a `reverifyLimit`, then the node will be marked as failing the audit and removed from containment mode.
+Otherwise, the node will pass the audit and will also be removed from containment mode.
+
+Additionally, contained nodes will be excluded from node selection.
+
+## Design
+New functionality needs to be added to the audit verifier, where upon receiving a pointer, the verifier should first iterate through the nodes listed in the pointer and check if they have a `contained` field set to true on their NodeDossier or not.
+
+For nodes that are not contained, the audit will continue as normal.
+
+For nodes that are contained, the verifier will ask the ContainmentDB for the hash of the erasure share expected from the node when it was originally audited.
+
+The verifier will then attempt to download that erasure share from the contained node.
+From there, the following cases can occur:
+- Passes audit ->  `pendingAudit` removed from the ContainmentDB  -> stats updated in overlaycache/satelliteDB
+- Fails audit -> `pendingAudit` is removed from the ContainmentDB -> stats updated in overlaycache/satelliteDB
+- Refuses to respond again -> the `pendingAudit`'s `reverifyCount` is incremented in the ContainmentDB ->
+    - -> If that count exceeds the reverify limit, then the node’s stats will be updated to reflect an audit failure and the node will be removed from the ContainmentDB
+    - -> If the count does not exceed the reverify limit, the `pendingAudit` will remain in the ContainmentDB
+
+```
+type pendingAudit struct {
+    nodeID       storj.NodeID
+    pieceID      storj.PieceID
+    pieceNum     int
+    stripeIndex  int
+    shareSize    int64
+    expectedShareHash []byte
+    reverifyCount   int // number of times reverify has been attempted
+}
+
+type ContainmentDB interface {
+  Contain(ctx context.Context, pendingAudit pendingAudit) error
```

Comment:
Why not `Put` rather than `Contain` as `Get` it's?

This is a question that can have a response like: "it's the common convention for Store interfaces"

Reply:
I wonder if we can/should combine Contain and ReverifyFail into a simple "IncrementFail" upsert

Reply:
That sounds like a good idea Bill! I originally had the interface with Get and Put, but Egon pointed out that that's susceptible to race conditions. I think it still is currently, but I think the `IncrementFail` idea could avoid that.

Reply:
I'm a little concerned though that `IncrementFail` could be confused with an "audit failure" since this isn't technically a failure yet. Does `IncrementPending` seem worse?

Reply:
sure!